Plugins

mariusandra opened the pull request:
Hey! I experimented with adding an abstract (background?) `painter` component that could be used to draw arbitrary things on the timeline: cell and row backgrounds, lines connecting items, webgl accelerated particles or videos of Kiefer Sutherland jumping into Christmas trees... In theory you could render anything without having to modify the timeline itself.

Here's a demo that uses it to display different background colors for rows:

<img width="1065" alt="screen shot 2017-06-28 at 10 11 21" src="https://user-images.githubusercontent.com/53387/27627553-7dee70ec-5bec-11e7-977f-5049d269939e.png">

It could be extended to add the `ranges` feature you mentioned in #74 or to fix #77. 

Basically the custom `painter` component would get a lot of internals from the timeline as props (`groupHeights`, `dimensionItems`, etc), and use this to do whatever is needed. 

Any thoughts on this approach?

signalwerk commented:
@mariusandra hey this is really cool! Wow! I reviewed this week some code of @meengit and he planed to send a PR for ranges. but with that method in place that's not necessary anymore. Instead he should have it as a separate npm as a "plugin" to load. 
So we should probably have two options for this prop. You can pass in a react-component or an array of react-components. So you can load multiple plug-ins

I would suggest the following API:

## untouched
* `canvasTimeStart`, `canvasTimeEnd` - start and end of the scrolling canvas in Unix timestamps
* `canvasWidth` - width of the scrolling canvas in pixels
* `visibleTimeStart`, `visibleTimeEnd` - start and end of the currently visible area
* `groups`, `items`, `keys` - groups, items and keys as passed to the timeline
* `dimensionItems` - an array of objects `{ id, dimensions: {...} }` describing positions of all the items
* `height`, `headerHeight` - height of the entire calendar (includes `headerHeight`) or just the header
* `timeSteps` - steps for displaying time

## new
* `dimensionGroups`- an array of objects `{ id, dimensions: {...} }` describing positions of all the groups
* `selected` - an array of Item ID of selected items. ((I swapped internally all to array handling > PR is coming next week. So there is no big difference between passed selected or internal selection handling. Therefore I would only give that one props to make it more uniform))

## drop
* ~~`lineCount` - number of groups~~ this could be done with `dimensionGroups.length`no?
* ~~`groupHeights`, `groupTops` - arrays of heights and tops for the groups~~ we have `dimensionGroups` ((this line is 2x in the readme.md)
* ~~`selectedItem`, `selected` - the selected item or controlled selected items array~~ 
 
I'm right now on an other project till mid next week so I can't invest too much time right now. But I try to keep up with the comments.

mariusandra commented:
Hey, thanks for the review!

Indeed after this, the `ranges` PR could and should be converted to a plugin. Your idea of allowing many plugins to be added at the same time sounds great. In fact, I'm thinking of redesigning the API to allow for this flexibility. My proposal is to allow something like this:

```jsx
<Timeline {...props}>
  <Ranges ranges={[/* 3pm to 5pm every day */, /* another range */]} />
  <RainbowBackground />
  <SomeOtherPlugin />
</Timeline>
```

All of these plugins would get cloned and passed these extra props. Similar to how I did it for [pigeon-maps](https://github.com/mariusandra/pigeon-maps).

With `v0.13.0` I deprecated passing the content for the left sidebar as children to the component, so there should be nothing in the way of this.

This might even pave the way for a fully composable timeline, where you could just do 

```jsx
import Timeline, { HorizontalLines, VerticalLines, TodayLine, Sidebar } from 'react-calendar-timeline'
import Ranges from 'react-calendar-timeline-ranges'

<Timeline {...timelineProps}>
  <Sidebar alignment='left' />
  <HorizontalLines />
  <VerticalLines />
  <TodayLine />
  <Ranges {...rangeProps} />
</Timeline>
```

... but that's to be thought of later. For sure before reaching 1.0 :).

For the other comments

* I'll wait for your PR on the `selectedItem` vs `selected` changes.
* Agreed about `lineCount` not being useful
* For merging `groupHeights` and `groupTops` into `dimensionGroups`, it makes some sense. However I would also like to avoid calls to create new objects just to pass to `painter`. I guess I could refactor the stacking functions to already return and use this combined object.

mariusandra commented:
Hey... and I did those API changes!

* The children of the component now get extra props... and thus I removed the `painter` prop.
* I merged `selectedIndex` and `selected` to just `selected`
* I removed lineCount

[Here's some code](https://github.com/mariusandra/react-calendar-timeline/blob/painter/demo/app/demo-plugins/index.js#L111)

I still didn't do anything in regards to `dimensionGroups`

signalwerk commented on README.md:
```diff
@@ -424,8 +421,23 @@ import containerResizeDetector from 'react-calendar-timeline/lib/resize-detector
+* `groups`, `items`, `keys` - groups, items and keys as passed to the timeline
+* `height`, `headerHeight` - height of the entire calendar (includes `headerHeight`) or just the header
+* `groupHeights`, `groupTops` - arrays of heights and tops for the groups
+* `groupHeights`, `groupTops` - arrays of heights and tops for the groups
```
this line is double

signalwerk commented:
I fucking like that new idea of yours;

```jsx
<Timeline {...props}>
  <Ranges ranges={[/* 3pm to 5pm every day */, /* another range */]} />
  <RainbowBackground />
  <SomeOtherPlugin />
</Timeline>
```

this is awesome. 

I would suggest to keep working on the  API. Stuff lie `dimensionGroups` is a bit work to rewrite our internal use. but it would also be much clearer. 

But yeah. Just merge it and we keep an eye on it. I send you tomorrow the PR of my [multiselect/multidrag feature]/(https://github.com/signalwerk/react-calendar-timeline/tree/feature/multiselect) and next week I try to adopt @meengit s ranges to a plugin. then I have a better view if the API fits for simple cases

mariusandra commented:
Cool! Thanks for the review! I'll add a notice in the README that the API for plugins will still change, so be prepared.

And I propose to move in the direction of unifying the internal and external APIs.... so we could also get rid of all the [pornographic methods](https://github.com/namespace-ee/react-calendar-timeline/blob/master/src/lib/Timeline.js#L1184-L1201) in Timeline.js that just render components (`this.items`, `this.verticalLines`, etc).

mariusandra commented:
I merged the code and will do a release with this (and the new header format) in the evening (CET). One more thing: I still need to write tests for this. That will also have to wait for the evening :).

signalwerk commented:
@mariusandra completely agree with your words! [eat your own dog food is always good](https://en.wikipedia.org/wiki/Eating_your_own_dog_food)

Don't hurry for the release. We are not in a rush. :)

mariusandra commented:
Well, it's merged and 0.15 is out! It think it's a good base for further improvements.

I also added [a simple test](https://github.com/namespace-ee/react-calendar-timeline/blob/master/src/lib/__tests__/index.js#L105-L139) to at least make sure the children get the right props. Once we settle on what those props should actually contain, we can extend it further.

Let the flood of new plugins begin! 🌊 😁
